File: backend/app/models/investment.py
import enum
import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, ForeignKey, Integer, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class InvestmentType(str, enum.Enum):
    stock = "stock"
    mutual_fund = "mutual_fund"
    fixed_deposit = "fixed_deposit"
    gold = "gold"
    crypto = "crypto"
    ppf = "ppf"
    nps = "nps"
    real_estate = "real_estate"


class Investment(Base):
    __tablename__ = "investments"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    platform_account_id: Mapped[int | None] = mapped_column(
        ForeignKey("platform_accounts.id", ondelete="SET NULL"), nullable=True, index=True
    )
    instrument_id: Mapped[int | None] = mapped_column(
        ForeignKey("instruments.id", ondelete="SET NULL"), nullable=True, index=True
    )
    type: Mapped[InvestmentType] = mapped_column(
        Enum(InvestmentType, name="investment_type"), index=True, nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    amount_invested: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    current_value: Mapped[float | None] = mapped_column(Numeric(14, 2), nullable=True)
    purchase_date: Mapped[date] = mapped_column(Date, nullable=False)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), onupdate=func.now(), nullable=True
    )

    # Stock / ETF  (ticker, exchange → on Instrument)
    quantity: Mapped[float | None] = mapped_column(Numeric(12, 4), nullable=True)
    avg_buy_price: Mapped[float | None] = mapped_column(Numeric(12, 2), nullable=True)

    # Mutual Fund  (fund_house → on Instrument)
    folio_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    units: Mapped[float | None] = mapped_column(Numeric(12, 4), nullable=True)
    nav_at_purchase: Mapped[float | None] = mapped_column(Numeric(12, 4), nullable=True)

    # Fixed Deposit
    bank_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    fd_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    interest_rate: Mapped[float | None] = mapped_column(Numeric(5, 2), nullable=True)
    tenure_months: Mapped[int | None] = mapped_column(Integer, nullable=True)
    maturity_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    maturity_amount: Mapped[float | None] = mapped_column(Numeric(14, 2), nullable=True)
    compounding: Mapped[str | None] = mapped_column(String(20), nullable=True)

    # Gold
    gold_form: Mapped[str | None] = mapped_column(String(30), nullable=True)
    weight_grams: Mapped[float | None] = mapped_column(Numeric(10, 3), nullable=True)
    purity: Mapped[str | None] = mapped_column(String(10), nullable=True)

    # Traceability — links back to Transaction.public_id (soft reference, no FK constraint)
    transaction_public_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True, index=True
    )

    platform_account = relationship("PlatformAccount", back_populates="investments")
    instrument = relationship("Instrument", foreign_keys=[instrument_id])

File: backend/app/models/investment.rb
# == Schema Information
#
# Table name: investments
#
#  id                    :bigint           not null, primary key
#  amount_invested       :decimal(14, 2)   not null
#  current_value         :decimal(14, 2)
#  folio_number          :string(50)
#  investment_type       :string           not null
#  name                  :string(255)      not null
#  notes                 :text
#  price                 :decimal(14, 4)
#  purchase_date         :date             not null
#  quantity              :decimal(12, 4)
#  trade_type            :string           default("buy"), not null
#  units                 :decimal(12, 4)
#  created_at            :datetime         not null
#  updated_at            :datetime         not null
#  order_id              :string(64)
#  platform_account_id   :bigint
#  trade_id              :string(64)
#  transaction_public_id :uuid
#  user_id               :bigint           not null
#  user_instrument_id    :bigint
#
# Indexes
#
#  index_investments_on_investment_type        (investment_type)
#  index_investments_on_order_id               (order_id)
#  index_investments_on_order_id_and_trade_id  (order_id,trade_id)
#  index_investments_on_platform_account_id    (platform_account_id)
#  index_investments_on_trade_id               (trade_id)
#  index_investments_on_trade_type             (trade_type)
#  index_investments_on_transaction_public_id  (transaction_public_id)
#  index_investments_on_user_id                (user_id)
#  index_investments_on_user_instrument_id     (user_instrument_id)
#
# Foreign Keys
#
#  fk_rails_...  (platform_account_id => platform_accounts.id) ON DELETE => nullify
#  fk_rails_...  (user_id => users.id) ON DELETE => cascade
#  fk_rails_...  (user_instrument_id => user_instruments.id) ON DELETE => nullify
#
class Investment < ApplicationRecord
  belongs_to :user
  belongs_to :platform_account, optional: true
  belongs_to :user_instrument,  optional: true
  has_one    :import_record, as: :importable, dependent: :nullify

  enum :investment_type, { stock: "stock", mutual_fund: "mutual_fund" }, validate: true
  enum :trade_type,      { buy:   "buy",   sell:        "sell" },        validate: true

  validates :name,            presence: true
  validates :amount_invested, presence: true, numericality: { greater_than_or_equal_to: 0 }
  validates :purchase_date,   presence: true

  scope :buys,  -> { where(trade_type: "buy") }
  scope :sells, -> { where(trade_type: "sell") }

  # Keep the cached Holding row in sync after every lot mutation. Runs in
  # Sidekiq so the request thread returns immediately. Bulk callers (CSV
  # import, API bulk endpoints) can set `Current.skip_holding_refresh = true`
  # and enqueue a single sweep at the end of the batch.
  after_save_commit    :enqueue_holding_refresh
  after_destroy_commit :enqueue_holding_refresh

  def enqueue_holding_refresh
    return if Current.skip_holding_refresh
    return unless user_instrument_id && platform_account_id
    Holdings::RefreshJob.perform_later(user_id, user_instrument_id, platform_account_id)
  end

  # Signed quantity: buy contributes +qty, sell contributes -qty. Used by holdings aggregation.
  def signed_quantity
    base = quantity || units || 0
    buy? ? base.to_f : -base.to_f
  end

  # Signed amount: buy contributes +invested, sell contributes -proceeds.
  # Net amount_invested across a position = total cash deployed minus cash returned.
  def signed_amount_invested
    buy? ? amount_invested.to_f : -amount_invested.to_f
  end
end
